{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/timp/miniconda3/envs/cpowgs/lib/python3.7/site-packages/rpy2/robjects/pandas2ri.py:14: FutureWarning: pandas.core.index is deprecated and will be removed in a future version.  The public classes are available in the top-level namespace.\n",
      "  from pandas.core.index import Index as PandasIndex\n"
     ]
    }
   ],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok - trying to use prokka - roary to make a pangenome\n",
    "\n",
    "Made a conda environment conda create --name cpowgs\n",
    "Then conda install -c conda-forge -c bioconda -c defaults prokka\n",
    "then\n",
    "conda config --add channels r\n",
    "conda config --add channels defaults\n",
    "conda config --add channels conda-forge\n",
    "conda config --add channels bioconda\n",
    "conda install roary\n",
    "then\n",
    "conda install jupyterlab (I know, wrong order)\n",
    "then\n",
    "conda install raxml\n",
    "then\n",
    "conda install snp-sites\n",
    "then\n",
    "conda install pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/d/Data/200404_trish/ecoli\n"
     ]
    }
   ],
   "source": [
    "%cd ~/Data/200404_trish/ecoli/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ok - clean up the pilon assmeblies so that the headers don't have _pilon_ over and over and over"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !sed -i 's/_pilon//g' pilon/*fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok - I need to pull down the genbank files for Enterobacter cloacae: https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/025/565/GCF_000025565.1_ASM2556v1/GCF_000025565.1_ASM2556v1_genomic.gbff.gz\n",
    "E. Coli: https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.gbff.gz\n",
    "and K. Pneumo: https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/240/185/GCF_000240185.1_ASM24018v2/GCF_000240185.1_ASM24018v2_genomic.gbff.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/d/Data/200404_trish/ecoli/refs\n",
      "--2020-04-05 15:43:59--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/240/185/GCF_000240185.1_ASM24018v2/GCF_000240185.1_ASM24018v2_genomic.gbff.gz\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.13, 2607:f220:41e:250::12\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.13|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 4011962 (3.8M) [application/x-gzip]\n",
      "Saving to: ‘GCF_000240185.1_ASM24018v2_genomic.gbff.gz’\n",
      "\n",
      "GCF_000240185.1_ASM 100%[===================>]   3.83M  --.-KB/s    in 0.1s    \n",
      "\n",
      "2020-04-05 15:43:59 (33.5 MB/s) - ‘GCF_000240185.1_ASM24018v2_genomic.gbff.gz’ saved [4011962/4011962]\n",
      "\n",
      "/mnt/d/Data/200404_trish/ecoli\n"
     ]
    }
   ],
   "source": [
    "%mkdir refs\n",
    "%cd refs\n",
    "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/240/185/GCF_000240185.1_ASM24018v2/GCF_000240185.1_ASM24018v2_genomic.gbff.gz\n",
    "!gunzip *gz\n",
    "%mv *gbff ecoli_ref.gb\n",
    "%cd .."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok - now prokka"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ecoli_100\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_100/ecoli_100.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_100/ecoli_100.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_102\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_102/ecoli_102.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_102/ecoli_102.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_173\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_173/ecoli_173.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_173/ecoli_173.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_174\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_174/ecoli_174.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_174/ecoli_174.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_271\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_271/ecoli_271.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_271/ecoli_271.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_387\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_387/ecoli_387.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_387/ecoli_387.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_402\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_402/ecoli_402.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_402/ecoli_402.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_421\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_421/ecoli_421.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_421/ecoli_421.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_485\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_485/ecoli_485.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_485/ecoli_485.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_50\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_50/ecoli_50.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_50/ecoli_50.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_503\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_503/ecoli_503.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_503/ecoli_503.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_516\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_516/ecoli_516.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_516/ecoli_516.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_52\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_52/ecoli_52.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_52/ecoli_52.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_526\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_526/ecoli_526.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_526/ecoli_526.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_56\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_56/ecoli_56.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_56/ecoli_56.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_594\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_594/ecoli_594.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_594/ecoli_594.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_630\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_630/ecoli_630.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_630/ecoli_630.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n",
      "ecoli_66\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_66/ecoli_66.gbf\n",
      "[tbl2asn-forever] Correcting dates in prok/ecoli_66/ecoli_66.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 05-APR-2020\n"
     ]
    }
   ],
   "source": [
    "%rm -R prok\n",
    "%rm -R gff\n",
    "%mkdir gff\n",
    "filelist=(glob.glob(\"pilon/*fasta\"))\n",
    "for myfile in filelist:\n",
    "    tag=re.split('[\\/\\.]', myfile)[1]\n",
    "    print(tag)\n",
    "    !prokka --quiet --protein refs/ecoli_ref.gb --outdir prok/$tag --genus Escherichia --usegenus --prefix $tag --locustag $tag $myfile \n",
    "    %cp prok/$tag/*gff gff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok - now let's try roary\n",
    "\n",
    "Roary says first https://github.com/microgenomics/tutorials/blob/master/pangenome.md\n",
    "you gotta organize the gff files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Use of uninitialized value in require at /home/timp/miniconda3/envs/cpowgs/lib/site_perl/5.26.2/x86_64-linux-thread-multi/Encode.pm line 61.\n",
      "\n",
      "Please cite Roary if you use any of the results it produces:\n",
      "    Andrew J. Page, Carla A. Cummins, Martin Hunt, Vanessa K. Wong, Sandra Reuter, Matthew T. G. Holden, Maria Fookes, Daniel Falush, Jacqueline A. Keane, Julian Parkhill,\n",
      "\t\"Roary: Rapid large-scale prokaryote pan genome analysis\", Bioinformatics, 2015 Nov 15;31(22):3691-3693\n",
      "    doi: http://doi.org/10.1093/bioinformatics/btv421\n",
      "\tPubmed: 26198102\n",
      "\n",
      "Use of uninitialized value in require at /home/timp/miniconda3/envs/cpowgs/lib/site_perl/5.26.2/x86_64-linux-thread-multi/Encode.pm line 61.\n",
      "Use of uninitialized value in require at (eval 931) line 1.\n"
     ]
    }
   ],
   "source": [
    "!roary -i 95 -f pan -e -n -p 8 gff/*gff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Counting snps from the multi-fa using snp-sites tool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "!snp-sites -v pan/core_gene_alignment.aln >pan/core_gene_alignment.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Parsed with column specification:\n",
      "cols(\n",
      "  .default = col_double(),\n",
      "  ID = \u001b[31mcol_character()\u001b[39m,\n",
      "  REF = \u001b[31mcol_character()\u001b[39m,\n",
      "  ALT = \u001b[31mcol_character()\u001b[39m,\n",
      "  QUAL = \u001b[31mcol_character()\u001b[39m,\n",
      "  FILTER = \u001b[31mcol_character()\u001b[39m,\n",
      "  INFO = \u001b[31mcol_character()\u001b[39m,\n",
      "  FORMAT = \u001b[31mcol_character()\u001b[39m\n",
      ")\n",
      "\n",
      "R[write to console]: See spec(...) for full column specifications.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R \n",
    "library(tidyverse)\n",
    "\n",
    "vcfs=read_tsv(\"pan/core_gene_alignment.vcf\", comment=\"##\")\n",
    "\n",
    "##ok - for each sample column, need to look at delta for all other sample columns, then sum to get that square on the table\n",
    "##Interesting names are names(vcfs)[10:22]\n",
    "\n",
    "comp=as_data_frame(t(combn(names(vcfs[10:22]),2))) %>%\n",
    "    mutate(val=colSums(abs(vcfs[V1]-vcfs[V2])))\n",
    "\n",
    "comp=rbind(comp, data_frame(V1=comp$V2, V2=comp$V1, val=comp$val))\n",
    "\n",
    "write_csv(spread(comp, V2, val), \"pan/core_gene_alignment_count.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok - that generated a pan genome and core alignment file which is already getting me close.  I then ran it in geneious, but let me take a guess and try to run in raxml\n",
    "some tips:\n",
    "-p is a random parsimony seed - to make sure analysis does the same thing\n",
    "-m GTRGAMMA - GTR GAMMA method\n",
    "-x is the rapid bootstrapping seed\n",
    "-N is the number of bootstrapops\n",
    "-f a says do rapid bootstrap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/d/Data/200404_trish/ecoli/raxml\n",
      "\n",
      "RAxML can't, parse the alignment file as phylip file \n",
      "it will now try to parse it as FASTA file\n",
      "\n",
      "\n",
      "\n",
      "Using BFGS method to optimize GTR rate parameters, to disable this specify \"--no-bfgs\" \n",
      "\n",
      "\n",
      "This is the RAxML Master Pthread\n",
      "\n",
      "This is RAxML Worker Pthread Number: 1\n",
      "\n",
      "This is RAxML Worker Pthread Number: 2\n",
      "\n",
      "This is RAxML Worker Pthread Number: 3\n",
      "\n",
      "This is RAxML Worker Pthread Number: 4\n",
      "\n",
      "This is RAxML Worker Pthread Number: 5\n",
      "\n",
      "\n",
      "This is RAxML version 8.2.12 released by Alexandros Stamatakis on May 2018.\n",
      "\n",
      "With greatly appreciated code contributions by:\n",
      "Andre Aberer      (HITS)\n",
      "Simon Berger      (HITS)\n",
      "Alexey Kozlov     (HITS)\n",
      "Kassian Kobert    (HITS)\n",
      "David Dao         (KIT and HITS)\n",
      "Sarah Lutteropp   (KIT and HITS)\n",
      "Nick Pattengale   (Sandia)\n",
      "Wayne Pfeiffer    (SDSC)\n",
      "Akifumi S. Tanabe (NRIFS)\n",
      "Charlie Taylor    (UF)\n",
      "\n",
      "\n",
      "Alignment has 23487 distinct alignment patterns\n",
      "\n",
      "Proportion of gaps and completely undetermined characters in this alignment: 3.33%\n",
      "\n",
      "RAxML rapid bootstrapping and subsequent ML search\n",
      "\n",
      "Using 1 distinct models/data partitions with joint branch length optimization\n",
      "\n",
      "\n",
      "\n",
      "Executing 100 rapid bootstrap inferences and thereafter a thorough ML search \n",
      "\n",
      "All free model parameters will be estimated by RAxML\n",
      "GAMMA model of rate heterogeneity, ML estimate of alpha-parameter\n",
      "\n",
      "GAMMA Model parameters will be estimated up to an accuracy of 0.1000000000 Log Likelihood units\n",
      "\n",
      "Partition: 0\n",
      "Alignment Patterns: 23487\n",
      "Name: No Name Provided\n",
      "DataType: DNA\n",
      "Substitution Matrix: GTR\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "RAxML was called as follows:\n",
      "\n",
      "raxmlHPC-PTHREADS -T 6 -f a -m GTRGAMMA -p 53243 -x 72341 -s ../pan/core_gene_alignment.aln -N 100 -n testrapid.tree \n",
      "\n",
      "\n",
      "\n",
      "Time for BS model parameter optimization 3.269915\n",
      "Bootstrap[0]: Time 5.990383 seconds, bootstrap likelihood -3926754.093899, best rearrangement setting 5\n",
      "Bootstrap[1]: Time 2.594129 seconds, bootstrap likelihood -3928344.406823, best rearrangement setting 11\n",
      "Bootstrap[2]: Time 2.201109 seconds, bootstrap likelihood -3929830.658544, best rearrangement setting 10\n",
      "Bootstrap[3]: Time 2.193909 seconds, bootstrap likelihood -3925449.096489, best rearrangement setting 13\n",
      "Bootstrap[4]: Time 2.471751 seconds, bootstrap likelihood -3933626.608091, best rearrangement setting 14\n",
      "Bootstrap[5]: Time 2.621192 seconds, bootstrap likelihood -3924567.574418, best rearrangement setting 6\n",
      "Bootstrap[6]: Time 2.574977 seconds, bootstrap likelihood -3938517.805997, best rearrangement setting 12\n",
      "Bootstrap[7]: Time 2.888768 seconds, bootstrap likelihood -3926562.789218, best rearrangement setting 9\n",
      "Bootstrap[8]: Time 2.944263 seconds, bootstrap likelihood -3921877.229478, best rearrangement setting 15\n",
      "Bootstrap[9]: Time 2.458238 seconds, bootstrap likelihood -3931092.483288, best rearrangement setting 14\n",
      "Bootstrap[10]: Time 2.509740 seconds, bootstrap likelihood -3934344.934374, best rearrangement setting 15\n",
      "Bootstrap[11]: Time 2.295050 seconds, bootstrap likelihood -3926953.351625, best rearrangement setting 15\n",
      "Bootstrap[12]: Time 2.082789 seconds, bootstrap likelihood -3927076.891322, best rearrangement setting 6\n",
      "Bootstrap[13]: Time 1.923030 seconds, bootstrap likelihood -3931618.092852, best rearrangement setting 9\n",
      "Bootstrap[14]: Time 1.814962 seconds, bootstrap likelihood -3926167.635513, best rearrangement setting 9\n",
      "Bootstrap[15]: Time 1.835096 seconds, bootstrap likelihood -3929328.616075, best rearrangement setting 13\n",
      "Bootstrap[16]: Time 1.985983 seconds, bootstrap likelihood -3928422.351104, best rearrangement setting 8\n",
      "Bootstrap[17]: Time 1.862779 seconds, bootstrap likelihood -3932671.824495, best rearrangement setting 11\n",
      "Bootstrap[18]: Time 2.063961 seconds, bootstrap likelihood -3930464.537353, best rearrangement setting 14\n",
      "Bootstrap[19]: Time 1.965196 seconds, bootstrap likelihood -3932644.373414, best rearrangement setting 11\n",
      "Bootstrap[20]: Time 2.339181 seconds, bootstrap likelihood -3929250.246482, best rearrangement setting 15\n",
      "Bootstrap[21]: Time 1.927495 seconds, bootstrap likelihood -3929622.123903, best rearrangement setting 11\n",
      "Bootstrap[22]: Time 2.126621 seconds, bootstrap likelihood -3924657.964478, best rearrangement setting 8\n",
      "Bootstrap[23]: Time 2.638263 seconds, bootstrap likelihood -3931729.257546, best rearrangement setting 10\n",
      "Bootstrap[24]: Time 2.376459 seconds, bootstrap likelihood -3931870.776877, best rearrangement setting 12\n",
      "Bootstrap[25]: Time 3.065628 seconds, bootstrap likelihood -3926134.854185, best rearrangement setting 5\n",
      "Bootstrap[26]: Time 2.268202 seconds, bootstrap likelihood -3928927.024643, best rearrangement setting 15\n",
      "Bootstrap[27]: Time 2.173276 seconds, bootstrap likelihood -3929052.396252, best rearrangement setting 15\n",
      "Bootstrap[28]: Time 2.195150 seconds, bootstrap likelihood -3926997.003966, best rearrangement setting 11\n",
      "Bootstrap[29]: Time 1.948748 seconds, bootstrap likelihood -3926492.706490, best rearrangement setting 15\n",
      "Bootstrap[30]: Time 1.970963 seconds, bootstrap likelihood -3930416.502716, best rearrangement setting 14\n",
      "Bootstrap[31]: Time 1.843643 seconds, bootstrap likelihood -3929383.322929, best rearrangement setting 12\n",
      "Bootstrap[32]: Time 1.799062 seconds, bootstrap likelihood -3927057.251415, best rearrangement setting 15\n",
      "Bootstrap[33]: Time 1.829376 seconds, bootstrap likelihood -3927917.329529, best rearrangement setting 5\n",
      "Bootstrap[34]: Time 1.839265 seconds, bootstrap likelihood -3927414.211214, best rearrangement setting 9\n",
      "Bootstrap[35]: Time 1.991379 seconds, bootstrap likelihood -3929548.530500, best rearrangement setting 11\n",
      "Bootstrap[36]: Time 1.950273 seconds, bootstrap likelihood -3930983.326719, best rearrangement setting 9\n",
      "Bootstrap[37]: Time 1.945148 seconds, bootstrap likelihood -3928735.073415, best rearrangement setting 9\n",
      "Bootstrap[38]: Time 1.886208 seconds, bootstrap likelihood -3930722.048696, best rearrangement setting 10\n",
      "Bootstrap[39]: Time 1.958553 seconds, bootstrap likelihood -3930438.729313, best rearrangement setting 7\n",
      "Bootstrap[40]: Time 1.947898 seconds, bootstrap likelihood -3922174.307349, best rearrangement setting 14\n",
      "Bootstrap[41]: Time 1.798089 seconds, bootstrap likelihood -3931347.448589, best rearrangement setting 11\n",
      "Bootstrap[42]: Time 1.870416 seconds, bootstrap likelihood -3929166.700160, best rearrangement setting 15\n",
      "Bootstrap[43]: Time 1.903560 seconds, bootstrap likelihood -3926945.806976, best rearrangement setting 13\n",
      "Bootstrap[44]: Time 1.911361 seconds, bootstrap likelihood -3927193.493677, best rearrangement setting 5\n",
      "Bootstrap[45]: Time 1.848281 seconds, bootstrap likelihood -3927852.998450, best rearrangement setting 10\n",
      "Bootstrap[46]: Time 3.931766 seconds, bootstrap likelihood -3929414.691081, best rearrangement setting 11\n",
      "Bootstrap[47]: Time 4.147934 seconds, bootstrap likelihood -3925572.691724, best rearrangement setting 7\n",
      "Bootstrap[48]: Time 1.922603 seconds, bootstrap likelihood -3931566.349184, best rearrangement setting 15\n",
      "Bootstrap[49]: Time 1.878616 seconds, bootstrap likelihood -3930972.062273, best rearrangement setting 5\n",
      "Bootstrap[50]: Time 2.094718 seconds, bootstrap likelihood -3931835.257789, best rearrangement setting 12\n",
      "Bootstrap[51]: Time 2.028363 seconds, bootstrap likelihood -3925993.430029, best rearrangement setting 12\n",
      "Bootstrap[52]: Time 1.959263 seconds, bootstrap likelihood -3933925.991183, best rearrangement setting 14\n",
      "Bootstrap[53]: Time 2.256564 seconds, bootstrap likelihood -3927218.970879, best rearrangement setting 9\n",
      "Bootstrap[54]: Time 2.029455 seconds, bootstrap likelihood -3936556.597534, best rearrangement setting 13\n",
      "Bootstrap[55]: Time 2.041455 seconds, bootstrap likelihood -3935072.000058, best rearrangement setting 12\n",
      "Bootstrap[56]: Time 2.216906 seconds, bootstrap likelihood -3926380.235388, best rearrangement setting 9\n",
      "Bootstrap[57]: Time 1.858401 seconds, bootstrap likelihood -3933585.584173, best rearrangement setting 5\n",
      "Bootstrap[58]: Time 1.813404 seconds, bootstrap likelihood -3932796.551312, best rearrangement setting 12\n",
      "Bootstrap[59]: Time 1.784989 seconds, bootstrap likelihood -3936282.593072, best rearrangement setting 14\n",
      "Bootstrap[60]: Time 1.900650 seconds, bootstrap likelihood -3929048.286626, best rearrangement setting 12\n",
      "Bootstrap[61]: Time 1.826776 seconds, bootstrap likelihood -3927349.962094, best rearrangement setting 10\n",
      "Bootstrap[62]: Time 1.812928 seconds, bootstrap likelihood -3924206.906424, best rearrangement setting 8\n",
      "Bootstrap[63]: Time 1.814013 seconds, bootstrap likelihood -3932189.098993, best rearrangement setting 10\n",
      "Bootstrap[64]: Time 1.818471 seconds, bootstrap likelihood -3929315.356223, best rearrangement setting 5\n",
      "Bootstrap[65]: Time 1.838362 seconds, bootstrap likelihood -3921113.002999, best rearrangement setting 8\n",
      "Bootstrap[66]: Time 1.759901 seconds, bootstrap likelihood -3930018.732257, best rearrangement setting 15\n",
      "Bootstrap[67]: Time 2.113988 seconds, bootstrap likelihood -3926430.668818, best rearrangement setting 9\n",
      "Bootstrap[68]: Time 2.232173 seconds, bootstrap likelihood -3928586.586040, best rearrangement setting 6\n",
      "Bootstrap[69]: Time 2.290044 seconds, bootstrap likelihood -3931948.888677, best rearrangement setting 9\n",
      "Bootstrap[70]: Time 2.135333 seconds, bootstrap likelihood -3930318.630310, best rearrangement setting 9\n",
      "Bootstrap[71]: Time 2.308412 seconds, bootstrap likelihood -3931104.317557, best rearrangement setting 11\n",
      "Bootstrap[72]: Time 2.488619 seconds, bootstrap likelihood -3934233.989987, best rearrangement setting 6\n",
      "Bootstrap[73]: Time 2.268538 seconds, bootstrap likelihood -3933691.670259, best rearrangement setting 7\n",
      "Bootstrap[74]: Time 2.061017 seconds, bootstrap likelihood -3928136.861663, best rearrangement setting 15\n",
      "Bootstrap[75]: Time 1.880657 seconds, bootstrap likelihood -3928595.743129, best rearrangement setting 7\n",
      "Bootstrap[76]: Time 1.879336 seconds, bootstrap likelihood -3931399.407881, best rearrangement setting 7\n",
      "Bootstrap[77]: Time 2.547510 seconds, bootstrap likelihood -3932061.018802, best rearrangement setting 13\n",
      "Bootstrap[78]: Time 2.109113 seconds, bootstrap likelihood -3934838.959809, best rearrangement setting 13\n",
      "Bootstrap[79]: Time 2.224745 seconds, bootstrap likelihood -3927466.234479, best rearrangement setting 6\n",
      "Bootstrap[80]: Time 4.160649 seconds, bootstrap likelihood -3935610.219475, best rearrangement setting 6\n",
      "Bootstrap[81]: Time 1.808563 seconds, bootstrap likelihood -3927436.792814, best rearrangement setting 11\n",
      "Bootstrap[82]: Time 3.722651 seconds, bootstrap likelihood -3923840.431955, best rearrangement setting 6\n",
      "Bootstrap[83]: Time 1.827789 seconds, bootstrap likelihood -3924198.355957, best rearrangement setting 9\n",
      "Bootstrap[84]: Time 1.839217 seconds, bootstrap likelihood -3929960.041564, best rearrangement setting 12\n",
      "Bootstrap[85]: Time 1.852090 seconds, bootstrap likelihood -3928800.913558, best rearrangement setting 9\n",
      "Bootstrap[86]: Time 1.808676 seconds, bootstrap likelihood -3929526.190923, best rearrangement setting 8\n",
      "Bootstrap[87]: Time 1.817431 seconds, bootstrap likelihood -3933961.619090, best rearrangement setting 9\n",
      "Bootstrap[88]: Time 1.846397 seconds, bootstrap likelihood -3926187.566585, best rearrangement setting 15\n",
      "Bootstrap[89]: Time 1.780268 seconds, bootstrap likelihood -3928698.752642, best rearrangement setting 8\n",
      "Bootstrap[90]: Time 2.025890 seconds, bootstrap likelihood -3933967.556243, best rearrangement setting 5\n",
      "Bootstrap[91]: Time 1.817608 seconds, bootstrap likelihood -3927686.296698, best rearrangement setting 6\n",
      "Bootstrap[92]: Time 1.993256 seconds, bootstrap likelihood -3927453.802216, best rearrangement setting 5\n",
      "Bootstrap[93]: Time 1.907607 seconds, bootstrap likelihood -3927316.190516, best rearrangement setting 8\n",
      "Bootstrap[94]: Time 1.757660 seconds, bootstrap likelihood -3932187.587607, best rearrangement setting 7\n",
      "Bootstrap[95]: Time 1.836153 seconds, bootstrap likelihood -3931634.692567, best rearrangement setting 12\n",
      "Bootstrap[96]: Time 1.838218 seconds, bootstrap likelihood -3925221.760122, best rearrangement setting 7\n",
      "Bootstrap[97]: Time 1.821651 seconds, bootstrap likelihood -3928977.451105, best rearrangement setting 10\n",
      "Bootstrap[98]: Time 1.831250 seconds, bootstrap likelihood -3931978.680967, best rearrangement setting 8\n",
      "Bootstrap[99]: Time 1.741066 seconds, bootstrap likelihood -3928858.393218, best rearrangement setting 5\n",
      "\n",
      "\n",
      "Overall Time for 100 Rapid Bootstraps 222.330325 seconds\n",
      "Average Time per Rapid Bootstrap 2.223303 seconds\n",
      "\n",
      "Starting ML Search ...\n",
      "\n",
      "Fast ML optimization finished\n",
      "\n",
      "Fast ML search Time: 201.361058 seconds\n",
      "\n",
      "Slow ML Search 0 Likelihood: -3928651.880277\n",
      "Slow ML Search 1 Likelihood: -3928651.880277\n",
      "Slow ML Search 2 Likelihood: -3928651.880259\n",
      "Slow ML Search 3 Likelihood: -3928651.880259\n",
      "Slow ML Search 4 Likelihood: -3928651.880285\n",
      "Slow ML Search 5 Likelihood: -3928651.880285\n",
      "Slow ML Search 6 Likelihood: -3928651.880260\n",
      "Slow ML Search 7 Likelihood: -3928651.880260\n",
      "Slow ML Search 8 Likelihood: -3928651.880298\n",
      "Slow ML Search 9 Likelihood: -3928651.880296\n",
      "Slow ML optimization finished\n",
      "\n",
      "Slow ML search Time: 130.089507 seconds\n",
      "Thorough ML search Time: 28.299973 seconds\n",
      "\n",
      "Final ML Optimization Likelihood: -3928651.880063\n",
      "\n",
      "Model Information:\n",
      "\n",
      "Model Parameters of Partition 0, Name: No Name Provided, Type of Data: DNA\n",
      "alpha: 0.020000\n",
      "Tree-Length: 0.118030\n",
      "rate A <-> C: 0.904967\n",
      "rate A <-> G: 4.202801\n",
      "rate A <-> T: 1.039622\n",
      "rate C <-> G: 0.606514\n",
      "rate C <-> T: 6.670158\n",
      "rate G <-> T: 1.000000\n",
      "\n",
      "freq pi(A): 0.235917\n",
      "freq pi(C): 0.248474\n",
      "freq pi(G): 0.277934\n",
      "freq pi(T): 0.237675\n",
      "\n",
      "\n",
      "ML search took 359.839195 secs or 0.099955 hours\n",
      "\n",
      "Combined Bootstrap and ML search took 582.177597 secs or 0.161716 hours\n",
      "\n",
      "Drawing Bootstrap Support Values on best-scoring ML tree ...\n",
      "\n",
      "\n",
      "\n",
      "Found 1 tree in File /mnt/d/Data/200404_trish/ecoli/raxml/RAxML_bestTree.testrapid.tree\n",
      "\n",
      "\n",
      "\n",
      "Found 1 tree in File /mnt/d/Data/200404_trish/ecoli/raxml/RAxML_bestTree.testrapid.tree\n",
      "\n",
      "Program execution info written to /mnt/d/Data/200404_trish/ecoli/raxml/RAxML_info.testrapid.tree\n",
      "All 100 bootstrapped trees written to: /mnt/d/Data/200404_trish/ecoli/raxml/RAxML_bootstrap.testrapid.tree\n",
      "\n",
      "Best-scoring ML tree written to: /mnt/d/Data/200404_trish/ecoli/raxml/RAxML_bestTree.testrapid.tree\n",
      "\n",
      "Best-scoring ML tree with support values written to: /mnt/d/Data/200404_trish/ecoli/raxml/RAxML_bipartitions.testrapid.tree\n",
      "\n",
      "Best-scoring ML tree with support values as branch labels written to: /mnt/d/Data/200404_trish/ecoli/raxml/RAxML_bipartitionsBranchLabels.testrapid.tree\n",
      "\n",
      "Overall execution time for full ML analysis: 582.209120 secs or 0.161725 hours or 0.006739 days\n",
      "\n",
      "/mnt/d/Data/200404_trish/ecoli\n"
     ]
    }
   ],
   "source": [
    "%rm -R raxml\n",
    "%mkdir raxml\n",
    "%cd raxml\n",
    "!raxmlHPC-PTHREADS -T 6 -f a -m GTRGAMMA -p 53243 -x 72341 -s ../pan/core_gene_alignment.aln -N 100 -n testrapid.tree\n",
    "%cd .."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
